File: Emulator/src/main/core/net/event/RegistryReadEvent.java
package core.net.event;

import core.game.world.GameWorld;
import core.net.IoReadEvent;
import core.net.IoSession;
import core.net.amsc.ManagementServerState;
import core.net.amsc.WorldCommunicator;
import core.net.producer.MSEventProducer;
import core.tools.Log;

import java.nio.ByteBuffer;

import static core.api.ContentAPIKt.log;

/**
 * The type Registry read event.
 */
public final class RegistryReadEvent extends IoReadEvent {

    private static final MSEventProducer PRODUCER = new MSEventProducer();

    /**
     * Instantiates a new Registry read event.
     *
     * @param session the session
     * @param buffer  the buffer
     */
    public RegistryReadEvent(IoSession session, ByteBuffer buffer) {
        super(session, buffer);
    }

    @Override
    public void read(IoSession session, ByteBuffer buffer) {
        int opcode = buffer.get() & 0xFF;
        switch (opcode) {
            case 0:
                WorldCommunicator.setState(ManagementServerState.NOT_AVAILABLE);
                log(this.getClass(), Log.ERR, "Failed registering world to AMS - [id=" + GameWorld.getSettings().getWorldId() + ", cause=World id out of bounds]!");
                break;
            case 1:
                session.setProducer(PRODUCER);
                WorldCommunicator.setState(ManagementServerState.AVAILABLE);
                break;
            case 2:
            case 3:
                WorldCommunicator.setState(ManagementServerState.NOT_AVAILABLE);
                break;
            default:

                break;
        }
    }

}

File: Emulator/src/main/core/net/event/RegistryReadEvent.kt
package core.net.event

import core.api.log
import core.game.world.GameWorld.settings
import core.net.IoReadEvent
import core.net.IoSession
import core.net.amsc.ManagementServerState
import core.net.amsc.WorldCommunicator.state
import core.net.producer.MSEventProducer
import core.tools.Log
import java.nio.ByteBuffer

/**
 * Handles world registry read events.
 *
 * @author Emperor
 */
class RegistryReadEvent(session: IoSession, buffer: ByteBuffer) : IoReadEvent(session, buffer) {
    override fun read(session: IoSession, buffer: ByteBuffer) {
        val opcode = buffer.get().toInt() and 0xFF
        when (opcode) {
            0 -> {
                state = ManagementServerState.NOT_AVAILABLE
                log(
                    this.javaClass,
                    Log.ERR,
                    "Failed registering world to AMS - [id=" + settings!!.worldId + ", cause=World id out of bounds]!"
                )
            }

            1 -> {
                session.setProducer(PRODUCER)
                state = ManagementServerState.AVAILABLE
            }

            2, 3 -> state = ManagementServerState.NOT_AVAILABLE
            else -> {}
        }
    }

    companion object {
        /**
         * The event producer.
         */
        private val PRODUCER = MSEventProducer()
    }
}